datafusion/core/src/physical_plan/joins/nested_loop_join.rs

```diff
@@ -0,0 +1,883 @@
+            // use the right partition
+            JoinType::Right => {
+                // if the partition of right is hash,
+                // and the right partition should be adjusted the column index for the right expr
+                adjust_right_output_partitioning(
+                    self.right.output_partitioning(),
+                    self.left.schema().fields.len(),
+                )
+            }
+            // use the right partition
+            JoinType::RightSemi | JoinType::RightAnti => self.right.output_partitioning(),
+        }
+    }
+
+    fn output_ordering(&self) -> Option<&[PhysicalSortExpr]> {
+        // no specified order for the output
+        None
+    }
+
+    fn required_input_distribution(&self) -> Vec<Distribution> {
+        distribution_from_join_type(&self.join_type)
+    }
+
+    fn equivalence_properties(&self) -> EquivalenceProperties {
+        let left_columns_len = self.left.schema().fields.len();
+        combine_join_equivalence_properties(
+            self.join_type,
+            self.left.equivalence_properties(),
+            self.right.equivalence_properties(),
+            left_columns_len,
+            &[], // empty join keys
+            self.schema(),
+        )
+    }
+
+    fn children(&self) -> Vec<Arc<dyn ExecutionPlan>> {
+        vec![self.left.clone(), self.right.clone()]
+    }
+
+    fn with_new_children(
+        self: Arc<Self>,
+        children: Vec<Arc<dyn ExecutionPlan>>,
+    ) -> Result<Arc<dyn ExecutionPlan>> {
+        Ok(Arc::new(NestedLoopJoinExec::try_new(
+            children[0].clone(),
+            children[1].clone(),
+            self.filter.clone(),
+            &self.join_type,
+        )?))
+    }
+
+    fn execute(
+        &self,
+        partition: usize,
+        context: Arc<TaskContext>,
+    ) -> Result<SendableRecordBatchStream> {
+        // if the distribution of left is `SinglePartition`, just need to collect the left one
+        let left_is_single_partition = self.is_single_partition_for_left();
+        // left side
+        let left_fut = if left_is_single_partition {
+            self.left_fut.once(|| {
+                // just one partition for the left side, and the first partition is all of data for left
+                load_left_specified_partition(0, self.left.clone(), context.clone())
+            })
+        } else {
+            // the distribution of left is not single partition, just need the specified partition for left
+            OnceFut::new(load_left_specified_partition(
+                partition,
+                self.left.clone(),
+                context.clone(),
+            ))
+        };
+        // right side
+        let right_side = if left_is_single_partition {
+            self.right.execute(partition, context)?
+        } else {
+            // the distribution of right is `SinglePartition`
+            self.right.execute(0, context)?
+        };
```

Comment:
Is it correct here ? For FullOut join, both the left input and right input are single partition now.

Reply:
Yes, if the join type is `Full Join`, the left and right is single partition.
```
    fn execute(
        &self,
        partition: usize,
        context: Arc<TaskContext>,
    ) -> Result<SendableRecordBatchStream> {
```
The parameter of `partition` must be `0`.

The parameter of `partition` is triggered by left side, if the left side is multi partition and have 4 partition, 
the value of `partition` will be `0`,`1`,`2`,`3`.

The implementation is right, but from your comments, I find it will make reviewer confused. I will add a new method to check if right  is single partition.